Add minor mode to enable/disable Magithub.

PierreTechoueyres opened the pull request:
This will allow you to do something like that in your initfile :

```
(add-hook 'magit-mode-hook 'turn-on-magithub)
```

And if course enabling/disabling Magithub when you want.

alphapapa commented:
Sorry, but I'm not sure I understand the point of this.  There's already a way to disable Magithub by pressing the backtick key.  And why would you want to disable it?  If there's a problem when it's enabled, it might make more sense to fix that problem than make a new minor mode to disable it.

PierreTechoueyres commented:
I'm sorry to be too vague.
Allowing the user to disable Magithub with an minor mode is just a side effect for what I expect.
Adding the minor mode and the function to enable it allow me to delay the loading of the file in my init.el.

I can do an `(add-hook 'magit-mode-hook #'turn-on-magithub)` and the loading of magithub.el will be delayed until I call magit.

But may be there is another way to achieve this.
I could do an `(add-hook 'magit-mode-hook '(lambda () (require 'magithub)))`. But this look too much as an dirty hack.

alphapapa commented:
If you just want to defer loading of the package, you can use `use-package` to load it like this:
```el
(use-package magithub
  :after magit)
```
If you're not already using `use-package`, I highly recommend it as a way to organize your init file and drastically improve startup time.  It's also made by the Emacs maintainer, John Wiegley.

PierreTechoueyres commented:
I know about use-package. This is a great piece of elisp even if I don't use it.
I still think the minor mode could be an improvement. But it's up to you.

alphapapa commented:
> But it's up to you.

Actually I'm just a bystander who made a PR myself and was curious about this one.  I'm not the author.  :)

vermiculus commented:
Sorry I've been absent the past few days; I've been visiting my girlfriend who lives a few hours away (and has also just given me a nasty cold).

Deferred loading aside (there are plenty of other ways to defer loading other than autoloaded commands), I think a minor mode to turn integrations on/off for *all* repositories would be useful. Keep in mind though that per-repository overrides should remain in effect – right now, we only have the ability to turn the integration *off* rather than explicitly turn it *on* for any given repository. In order to merge this, I'd have to double-check (or write) a means to explicitly enable the repository. This check should be done in `magithub-usable-p`.

vermiculus commented on magithub.el:
```diff
@@ -71,10 +76,34 @@
+  :lighter  magithub-mode-lighter
+  :require 'magithub
+  :keymap  'magithub-mode-map
+  (or (derived-mode-p 'magit-mode)
```
Would it make sense to define the minor mode in such a way that we could enable it with `(turn-on-magithub-mode)` as a top-level form?

vermiculus commented on magithub.el:
```diff
@@ -71,10 +76,34 @@
              (?& "Request a feature or report a bug" magithub--meta-new-issue)
              (?h "Ask for help on Gitter" magithub--meta-help)))
 
-(magit-define-popup-action 'magit-dispatch-popup
```
I don't want to remove the action from the dispatch popup. That shouldn't be dependent on the minor mode at all. See magit/magit#2779.

PierreTechoueyres replied:
Sorry, I don't follow you here. See my comment at end.

PierreTechoueyres commented on magithub.el:
```diff
@@ -71,10 +76,34 @@
+  (or (derived-mode-p 'magit-mode)
+      (error "This mode only makes sense with magit"))
+  (if magithub-mode
+      (progn
```
Oups, bad copy / paste ...

PierreTechoueyres commented:
Sorry for the long delay.
I think there are some misunderstanding between our expectations about this PR.
I'll try to explain what I had expected.
First I wanted to be able to delay the loading of magithub until I start using magit. So I added the minor mode, the associated mode key-map and the turn-on function.
For the minor mode, I expected that when it was disabled the menu entry and the key-map where also removed. IMHO the minor mode should remove magit's associated  menu  entry.
What I've missed is that magithub activation is conditioned by many conditions, so I should modify the minor mode to take account of `magithub-usable-p`
If you're agree with theses explanations I could try to improve the PR.

PierreTechoueyres commented on magithub.el:
```diff
@@ -71,10 +76,34 @@
+  :require 'magithub
+  :keymap  'magithub-mode-map
+  (or (derived-mode-p 'magit-mode)
+      (error "This mode only makes sense with magit"))
```
`user-error` seem more standard.

vermiculus commented:
If the end goal is to delay loading magithub until it's actually needed, why not just introduce `;;;###autoload` cookies to all the interactive functions? I'm still hesitant to think about magithub as a magit minor mode rather than as a magit extension.

PierreTechoueyres commented:
My first goal was to delay the loading and I've used
```emacs-lisp
(eval-after-load "magit" '(require 'magithub))
```
Then I wanted to be able to remove the magit-menu reference for magithub (for various raisons) on certain repo.

You already allow to enable it only on certain repo but the menu still remains (which I wanted to avoid).

Maybe the minor-mode is just overkill, and an autoloaded function `magithub-setup-magit` to setup the menu and the keys when `magithub-usable-p` is true could be sufficient ?
Then I could use `(add-hook 'magit-mode-setup-hook #'magithub-setup-magit)`
